src/cert_fetcher.rs

```diff
@@ -45,27 +47,42 @@ pub fn new(cfg: &config::Config, cert_manager: Arc<SecretManager>) -> Arc<dyn Ce
 struct CertFetcherImpl {
     proxy_mode: ProxyMode,
     local_node: Option<String>,
-    tx: mpsc::Sender<Identity>,
+    tx: mpsc::Sender<Action>,
+}
+
+#[derive(Debug)]
+enum Action {
+    Fetch(Identity),
+    Forget(Identity),
 }
 
 impl CertFetcherImpl {
     fn new(cfg: &config::Config, cert_manager: Arc<SecretManager>) -> Self {
-        let (tx, mut rx) = mpsc::channel::<Identity>(256);
+        let (tx, mut rx) = mpsc::channel::<Action>(256);
 
         // Spawn a task for handling the pre-fetch requests asynchronously.
         tokio::spawn(async move {
-            while let Some(workload_identity) = rx.recv().await {
-                match cert_manager
-                    .fetch_certificate_pri(&workload_identity, Warmup)
-                    .await
-                {
-                    Ok(_) => debug!("prefetched cert for {:?}", workload_identity.to_string()),
-                    Err(e) => error!(
-                        "unable to prefetch cert for {:?}, skipping, {:?}",
-                        workload_identity.to_string(),
-                        e
-                    ),
-                }
+            while let Some(action) = rx.recv().await {
+                match action {
+                    Action::Fetch(workload_identity) => {
+                        match cert_manager
+                            .fetch_certificate_pri(&workload_identity, Warmup)
+                            .await
+                        {
+                            Ok(_) => {
+                                debug!("prefetched cert for {:?}", workload_identity.to_string())
+                            }
+                            Err(e) => error!(
+                                "unable to prefetch cert for {:?}, skipping, {:?}",
+                                workload_identity.to_string(),
+                                e
+                            ),
+                        }
+                    }
+                    Action::Forget(workload_identity) => {
+                        cert_manager.forget_certificate(&workload_identity).await;
```

Comment:
I kind of wonder if we are going to get in a state where we flip flop. Imagine I have a single node cluster and a 1 replica deployment and it is deleted+restarted.

Or maybe the manager handles this already

Reply:
I think if the only pod with some SA on a node is flapping then we would do a lot of fetches and forgets in this implementation.